src/fs_watcher.rs

```diff
@@ -0,0 +1,135 @@
+use crate::Dst;
+use crate::DstSender;
+use crate::EndpointMeta;
+use crate::Endpoints;
+use inotify::{Event, EventMask, Inotify, WatchMask};
+use inotify_sys as ffi;
+use serde_json;
+use serde_yaml;
+use std::error::Error;
+use std::ffi::OsString;
+use std::mem;
+use std::{fs, path::PathBuf};
+use tokio::stream::StreamExt;
+
+const EVENT_BUF_SZ: usize =
+    mem::size_of::<ffi::inotify_event>() + (libc::FILENAME_MAX as usize) + 1;
+
+#[derive(Debug)]
+pub struct FsWatcher {
+    endpoints_dir: PathBuf,
+    dst_sender: DstSender,
+}
+
+#[derive(Debug)]
+enum FileType {
+    Yaml,
+    Json,
+}
+
+#[derive(Debug)]
+pub struct FsWatcherError {
+    reason: &'static str,
+}
+
+macro_rules! fs_watcher_error {
+    ($reason:expr) => {{
+        return Err(Box::new(FsWatcherError { reason: $reason }));
+    }};
+}
+
+// === impl ParseError ===
+
+impl std::fmt::Display for FsWatcherError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        std::fmt::Display::fmt(&self.reason, f)
+    }
+}
+
+impl Error for FsWatcherError {}
+
+impl FsWatcher {
+    pub fn new(endpoints_dir: PathBuf, dst_sender: DstSender) -> Self {
+        Self {
+            endpoints_dir,
+            dst_sender,
+        }
+    }
+
+    fn parse_dst(
+        file_name: &str,
+    ) -> Result<(Dst, FileType), Box<dyn Error + Send + Sync + 'static>> {
+        let mut parts = file_name.rsplitn(2, ".");
+        match (parts.next(), parts.next()) {
+            (Some(ext), Some(name)) => {
+                let dst: Dst = name.parse()?;
+                let ft = match ext {
+                    "yaml" => FileType::Yaml,
+                    "json" => FileType::Json,
+                    _ => fs_watcher_error!("invalid file ext"),
```

Comment:
It might be worth  changing the `&'static str` in the error type into a dynamic `String` so that we can include what the invalid extension was? I don't think this code is performance-critical enough for avoiding the string allocation to be important...but if it was, we could use[`Cow<'static, str>`](https://doc.rust-lang.org/stable/std/borrow/enum.Cow.html) to avoid the allocation when there is no dynamic formatting to do.